sdk/azidentity/device_code_credential.go

```diff
@@ -0,0 +1,100 @@
+	DeviceCode      string `json:"device_code"`      // Device code returned by the service
+	VerificationURL string `json:"verification_uri"` // Verification URL where the user must navigate to authenticate using the device code and credentials.
+	Interval        int64  `json:"interval"`         // Polling interval time to check for completion of authentication flow.
+	Message         string `json:"message"`          // User friendly text response that can be used for display purpose.
+}
+
+// DeviceCodeCredential authenticates a user using the device code flow, and provides access tokens for that user account.
+// For more information on the device code authentication flow see https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-device-code
+type DeviceCodeCredential struct {
+	client       *aadIdentityClient
+	tenantID     string       // Gets the Azure Active Directory tenant (directory) Id of the service principal
+	clientID     string       // Gets the client (application) ID of the service principal
+	callback     func(string) // Sends the user a message with a verification URL and device code to sign in to the login server
+	refreshToken string       // Gets the refresh token sent from the service and will be used to retreive new access tokens after the initial request for a token
+}
+
+// NewDeviceCodeCredential constructs a new DeviceCodeCredential with the details needed to authenticate against Azure Active Directory with a device code.
+// tenantID: The Azure Active Directory tenant (directory) Id of the service principal.
+// clientID: The client (application) ID of the service principal.
+// callback: The callback function used to send the login message back to the user
+// options: allow to configure the management of the requests sent to the Azure Active Directory service.
+func NewDeviceCodeCredential(tenantID string, clientID string, callback func(string), options *TokenCredentialOptions) *DeviceCodeCredential {
+	return &DeviceCodeCredential{tenantID: tenantID, clientID: clientID, callback: callback, client: newAADIdentityClient(options)}
+}
+
+// GetToken obtains a token from the Azure Active Directory service, following the device code authentication
+// flow. This function first requests a device code and requests that the user login to continue authenticating.
+// This function will keep polling the service for a token meanwhile the user logs.
+// - scopes: The list of scopes for which the token will have access.
+// - ctx: controlling the request lifetime.
+// Returns an AccessToken which can be used to authenticate service client calls.
+func (c *DeviceCodeCredential) GetToken(ctx context.Context, opts azcore.TokenRequestOptions) (*azcore.AccessToken, error) {
+	if len(c.refreshToken) != 0 {
+		tk, err := c.client.refreshAccessToken(ctx, c.tenantID, c.clientID, "", c.refreshToken, opts.Scopes)
+		if tk == nil { // this would only happen if there was an unmarshal error or a failure to parse ExpiresOn
```

Comment:
This should be `err != nil`, we should only predicate failure on a non-nil error.

Reply:
Ok, got it 👍